### PR TITLE
*: asynchronous pd client, take 4

### PR DESCRIPTION
**Pull request by overvenus**
This is the fourth part of using an asynchronous PD client. It contains the following changes:

 - [X] Refactor PdWorker as AsyncWorker. (Address https://github.com/pingcap/tikv/pull/1712#issuecomment-290102625) 

Done:
 - [x] Asynchronous retry. #1712 
 - [x] Refactor trait `PdClient`.  #1745 
 - [x] Add an AsyncWorker, it runs a tokio core. (Address https://github.com/pingcap/tikv/pull/1712#issuecomment-290088819) #1752 

Next PRs:
 - [ ] Add some metrics about asynchronous tasks.
 - [ ] Try to replace boxed `Future` with `impl Future` . (Address https://github.com/pingcap/tikv/pull/1712#issuecomment-290329925)

### CODE REVIEW

**Comment by siddontang** [comment]
LGTM

PTAL @BusyJay @disksing

File: src/raftstore/store/worker/pd.rs [hunk 1]
```diff
@@ -97,183 +98,194 @@ impl<T: PdClient> Runner<T> {
...
                           peer.get_id(),
                           pd_region);
-                    PD_VALIDATE_PEER_COUNTER_VEC.with_label_values(&["peer stale"]).inc();
-                    self.send_destroy_peer_message(local_region, peer, pd_region);
-                    return;
+                    PD_VALIDATE_PEER_COUNTER_VEC.with_label_values(&["peer valid"]).inc();
+                    Ok(())
+                }
+                Ok(None) => {
+                    // split region has not yet report to pd.
```
Comment (andelf) [inline src/raftstore/store/worker/pd.rs]: reported
splitted

File: src/raftstore/store/worker/pd.rs [hunk 2]
```diff
@@ -97,195 +98,209 @@ impl<T: PdClient> Runner<T> {
...
+                            let req = new_transfer_leader_request(transfer_leader.take_peer());
+                            send_admin_request(&ch, region, peer, req)
+                        }
+                        Ok(())
+                    }
+                    Err(e) => {
+                        debug!("[region {}] failed to send heartbeat: {:?}",
+                               region.get_id(),
+                               e);
+                        Err(())
```
Comment (BusyJay) [inline src/raftstore/store/worker/pd.rs]: `Err(())` doesn't make sense, return `Ok(())` instead.

File: src/raftstore/store/worker/pd.rs [hunk 3]
```diff
@@ -316,3 +331,44 @@ fn new_transfer_leader_request(peer: metapb::Peer) -> AdminRequest {
     req.mut_transfer_leader().set_peer(peer);
     req
 }
+
+fn send_admin_request(ch: &SendCh<Msg>,
```
Comment (BusyJay) [inline src/raftstore/store/worker/pd.rs]: Move instead.

File: src/raftstore/store/worker/pd.rs [hunk 4]
```diff
@@ -316,3 +331,44 @@ fn new_transfer_leader_request(peer: metapb::Peer) -> AdminRequest {
...
+    if let Err(e) = ch.try_send(Msg::new_raft_cmd(req, Box::new(|_| {}))) {
+        error!("[region {}] send {:?} request err {:?}",
+               region_id,
+               cmd_type,
+               e);
+    }
+}
+
+// send a raft message to destroy the specified stale peer
+fn send_destroy_peer_message(ch: &SendCh<Msg>,
```
Comment (BusyJay) [inline src/raftstore/store/worker/pd.rs]: Move instead.

**Comment by siddontang** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @BusyJay @andelf

File: src/raftstore/store/worker/pd.rs [hunk 5]
```diff
@@ -97,195 +98,204 @@ impl<T: PdClient> Runner<T> {
...
+                        // Just add a log here for this situation.
+                        error!("[region {}] {} the local region epoch: {:?} is greater the \
+                                region epoch in PD: {:?}. Something is wrong!",
+                               local_region.get_id(),
+                               peer.get_id(),
+                               local_region.get_region_epoch(),
+                               pd_region.get_region_epoch());
+                        PD_VALIDATE_PEER_COUNTER_VEC.with_label_values(&["region epoch error"])
+                            .inc();
+                        return Err(());
```
Comment (BusyJay) [inline src/raftstore/store/worker/pd.rs]: return `Ok(())` instead.

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**